### PR TITLE
[imgui] Add test-engine feature

### PR DESCRIPTION
**Pull request by RT2Code**
Add test-engine feature to the imgui port: https://github.com/ocornut/imgui_test_engine

I made it a feature because they both depend on each other to build, that's why it is a feature instead of a new port.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version.~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/imgui/portfile.cmake [hunk 1]
```diff
@@ -57,6 +58,20 @@ if ("libigl-imgui" IN_LIST FEATURES)
...
+if ("test-engine" IN_LIST FEATURES)
+    vcpkg_from_github(
+        OUT_SOURCE_PATH TEST_ENGINE_SOURCE_PATH
+        REPO ocornut/imgui_test_engine
+        REF "v${VERSION}"
+        SHA512 809b06076fbeb544cd9544020c336f943f05bc5772df183c94dbdd5057d9b7b8718c72e5b908205cbef06c511b01f7a8e706a23aa668ca9fd12e891ef8ffb48e
+        HEAD_REF master
+    )
+
+    file(COPY "${TEST_ENGINE_SOURCE_PATH}/imgui_test_engine/" DESTINATION "${SOURCE_PATH}/test-engine")
```
Comment (dg0yt) [inline ports/imgui/portfile.cmake:70]:
```suggestion
    file(REMOVE_RECURSE "${SOURCE_PATH}/test-engine")
    file(COPY "${TEST_ENGINE_SOURCE_PATH}/imgui_test_engine/" DESTINATION "${SOURCE_PATH}/test-engine")
```
"Editable" build do not remove the sources.

File: ports/imgui/portfile.cmake [hunk 2]
```diff
@@ -57,6 +58,20 @@ if ("libigl-imgui" IN_LIST FEATURES)
...
+    vcpkg_replace_string("${SOURCE_PATH}/imconfig.h" "#pragma once" "#pragma once\n\n#include \"imgui_te_imconfig.h\"")
```
Comment (dg0yt) [inline ports/imgui/portfile.cmake:71]:
```suggestion
    vcpkg_replace_string("${SOURCE_PATH}/imconfig.h" "#pragma once" "#pragma  once\n\n#include \"imgui_te_imconfig.h\"")
```
Again, handle editable builds.

File: ports/imgui/portfile.cmake [hunk 3]
```diff
@@ -80,4 +95,8 @@ endif()
 vcpkg_copy_pdbs()
 vcpkg_cmake_config_fixup()
 
-vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE.txt")
+if ("test-engine" IN_LIST FEATURES)
+    vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/test-engine/LICENSE.txt")
+else()
+    vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE.txt")
+endif()
```
Comment (dg0yt) [inline ports/imgui/portfile.cmake:102]: This is surprising.  
If the licenses are the same, if/else is not needed.
If they are different, the FILE_LIST should perhaps be extended for test-engine.

File: ports/imgui/portfile.cmake [hunk 4]
```diff
@@ -57,6 +58,20 @@ if ("libigl-imgui" IN_LIST FEATURES)
     file(INSTALL "${IMGUI_FONTS_DROID_SANS_H}" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
 endif()
 
+if ("test-engine" IN_LIST FEATURES)
+    vcpkg_from_github(
+        OUT_SOURCE_PATH TEST_ENGINE_SOURCE_PATH
+        REPO ocornut/imgui_test_engine
+        REF "v${VERSION}"
```
Comment (dg0yt) [inline ports/imgui/portfile.cmake:65]: Do we need a test port to activate this feature?
It is easy to forget the test, or just to forget to update the SHA512 for a optional download.

File: ports/imgui/portfile.cmake (hunk 3 above)
Comment (RT2Code) [reply ports/imgui/portfile.cmake]: The licenses are different: imgui uses MIT, but the test engine has its own license. This is an oversight on my part, both licenses should indeed be installed with the test-engine feature, I fixed it.

File: ports/imgui/portfile.cmake (hunk 4 above)
Comment (RT2Code) [reply ports/imgui/portfile.cmake:65]: I didn't know that test ports existed, but I suppose it would be better to have one here. I added it.

**Comment by dg0yt** [comment]
Now we see how useful test ports and executables are...
With test-engine in imgui, there are two problems downstream:
- stb definitions in test engine duplicate the definitions in downstream apps.
- stb definitions in test engine are not relocatable.

Excerpt from milerius-sfml-imgui:
~~~
/mnt/vcpkg-ci/installed/x64-linux/include/stb_image_write.h:1617: multiple definition of `stbi_write_jpg'; /mnt/vcpkg-ci/installed/x64-linux/debug/lib/libimguid.a(imgui_capture_tool.cpp.o):/mnt/vcpkg-ci/b/imgui/src/.6-docking-82d03bc653.clean/test-engine/thirdparty/stb/imstb_image_write.h:1519: first defined here
/usr/bin/ld: /mnt/vcpkg-ci/installed/x64-linux/debug/lib/libsfml-graphics-s-d.a(ImageLoader.cpp.o): warning: relocation against `stbi_write_force_png_filter' in read-only section `.text'
/usr/bin/ld: /mnt/vcpkg-ci/installed/x64-linux/debug/lib/libsfml-graphics-s-d.a(ImageLoader.cpp.o): relocation R_X86_64_PC32 against symbol `stbi_write_tga_with_rle' can not be used when making a shared object; recompile with -fPIC
~~~

IMO there should be no stb definition in any lib (unless there would be a *single* stb lib) - that's how stb is designed.

**Comment by RT2Code** [comment]
I managed to make it build on all platforms. I disabled uwp support because the test-engine library uses some CRT functions which aren't supported on this platform like _wpopen. I also replaced the embedded stb lib with the one from vcpkg to better conform to the vcpkg maintainer guidelines.

One thing that I'm really unhappy with is the way that the std implementation is managed. The only way I found to make it build everywhere was to disable it for static builds, but keep it enabled for dynamic builds, which is unconsistent.

Wouldn't it be better if the stb port compiled it to a static library instead of only providing the headers and leaving the responsability of handling the implementation to the user? The macro would need to be defined only once, in the port build, and then it would act like any other static library. No multiple definitions, no need to define any macro to get the implementation.

**Review by WangWeiLin-MV — DISMISSED** [review]
~The port usage and features test pass with the following triplets:~
* ~x64-linux~
* ~x64-windows~

Test invalid since PR updated.

**Comment by dg0yt** [comment]
> Wouldn't it be better if the stb port compiled it to a static library instead of only providing the headers and leaving the responsability of handling the implementation to the user? The macro would need to be defined only once, in the port build, and then it would act like any other static library. No multiple definitions, no need to define any macro to get the implementation.

That is also what I think. There are several such claim to be header only, but actually only avoid building a proper static lib. 
However I don't want to open this can of worms for packages which document another form of use.

**Comment by RT2Code** [comment]
I spent more time studying the stb libs and reconsidered what I said. Those libs have many important compile time options, and some are designed to build with static symbols to allow for multiple configurations in the same project. If we build them into a static library, we remove those options for the user, so I now think the current way vcpkg handle stb is the best.

That being said, I have found a better way to fix the multiple definition issue. As I said, some stb libs provide the option to make the symbols static instead of public in the compilation unit, and it's perfect to prevent any clash with other libraries using the same stb headers. I just added the STB_IMAGE_WRITE_STATIC macro before std_image_write.h inclusion, and now everything builds fine on every platform with no ugly hacks.

**Review by WangWeiLin-MV — DISMISSED** [review]
The port usage tests pass with the following triplets:
* x64-linux
* x64-windows

**Comment by RT2Code** [comment]
Sorry, I changed it to draft because I want to try a different approach to add this feature which could be better. I'll make a second PR and I'll explain the difference between the two versions, so you can choose the one you think is the most pertinent.

**Comment by BillyONeal** [comment]
In general, ports are not expected to build tests. Can you explain what a vcpkg customer does with this option? (This might be entirely fine, I just want to understand that it is :) )

**Comment by RT2Code** [comment]
The test-engine library is an official component of imgui composed of two parts :
- The test-engine itself, a test/automation system for imgui.
- The test-suite, a collection of tests for imgui written with the test-engine.

The latter is not relevant for vcpkg, because as you said, ports are not supposed to build tests, that's why I ignore it in this port. But the test-engine can be very useful for a variety of applications. It's akin to libraries like google-test or catch2, it can be used to write your own tests for imgui, or for automation purposes. As it can simulate user inputs, it can also be used to write some kind of live tutorials, and it is also able to export the result to videos/gifs to showcase it. There's also a performance tool, and that's the feature that is lacking in the current PR that I want to try to add, but it's not easy because of a circular dependency with implot. 

In a nutshell, the test-engine is a very powerful set of features for imgui, and it really should be available in vcpkg.

**Comment by RT2Code** [comment]
Ultimately, I was unable to make the new PR work [for reasons explained here](https://github.com/microsoft/vcpkg/pull/38961#issuecomment-2133714138), so, except if someone knows a way to fix it, I guess this one will have to do.

File: versions/i-/imgui.json [hunk 5]
```diff
@@ -1,7 +1,12 @@
 {
   "versions": [
     {
-      "git-tree": "3c5ccb0267846fbcf5153a375c6e05c65ad3c7ab",
+      "git-tree": "af958fee35b0e48dd0397a6f807e28a3f7bb99ce",
+      "version": "1.90.6",
+      "port-version": 1
+    },
+    {
+      "git-tree": "64d9267db2a90b61bc06b597c63a06611d610caa",
```
Comment (vicroms) [inline versions/i-/imgui.json:9]: only the new version should be added please revert the change to `1.90.6`

**Review by vicroms — CHANGES_REQUESTED** [review]

**Review by vicroms — APPROVED** [review]
